Can't apply `automock` to traits with default method implementions marked `#[inline]`

The following code fails to compile:

```rust
#[mockall::automock]
trait Foo {
    #[inline]
    fn foo() -> i32 {
        1
    }
}

fn main() {}
```

With the error:

```
error[E0518]: attribute should be applied to function or closure
 --> src/main.rs:3:5
  |
1 | #[mockall::automock]
  | -------------------- not a function or closure
2 | trait Foo {
3 |     #[inline]
  |     ^^^^^^^^^
```

The issue is that the default method implementation is marked `#[inline]`. The same problem occurs for `#[inline(always|never)]`.

mockall: 0.12.1
rustc: 1.76.0

---

I also checked `#[cold]` and `#[must_use]`. `#[cold]` causes a warning to be emitted that will become an error in the future:

```
warning: attribute should be applied to a function definition
 --> src/main.rs:3:5
  |
1 | #[mockall::automock]
  | -------------------- not a function definition
2 | trait Foo {
3 |     #[cold]
  |     ^^^^^^^
  |
  = warning: this was previously accepted by the compiler but is being phased out; it will become a hard error in a future release!
  = note: `#[warn(unused_attributes)]` on by default
```

`#[must_use]` causes a warning, but as of Rust 1.76.0, it's not scheduled to become a hard error (edit: oops, pasted in the wrong warning originally):

```
warning: `#[must_use]` has no effect when applied to a module
 --> src/main.rs:3:5
  |
3 |     #[must_use]
  |     ^^^^^^^^^^^
  |
  = note: `#[warn(unused_attributes)]` on by default
```

There may be other attributes that cause problems, but these were all I tested.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -5,6 +5,11 @@ This project adheres to [Semantic Versioning](http://semver.org/).
 
 ## [ Unreleased ] - ReleaseDate
 
+### Added
+
+- Add the ability to mock methods that use `#[inline]` or `#[cold]`.
+  ([#555](https://github.com/asomers/mockall/pull/555))
+
 ### Changed
 
 - Raised MSRV to 1.70.0 to remove `lazy_static` dependency

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -7,7 +7,8 @@ This project adheres to [Semantic Versioning](http://semver.org/).
 
 ### Added
 
-- Add the ability to mock methods that use `#[inline]` or `#[cold]`.
+- Add the ability to mock methods that use `#[inline]` or `#[cold]`, and
+  methods or traits that use `#[must_use]`.
   ([#555](https://github.com/asomers/mockall/pull/555))
 
 ### Changed

--- a/mockall/tests/automock_inline.rs
+++ b/mockall/tests/automock_inline.rs
@@ -1,0 +1,52 @@
+// vim: tw=80
+//! Mockall should ignore and not emit attributes like "inline" that affect
+//! code generation.
+#![deny(warnings)]
+
+use mockall::*;
+
+pub struct Foo {}
+
+#[automock]
+impl Foo {
+    #[inline]
+    pub fn foo(&self) -> i32 {unimplemented!()}
+    #[inline]
+    pub fn bar() -> i32 {unimplemented!()}
+    #[cold]
+    pub fn baz(&self) -> i32 {unimplemented!()}
+    #[cold]
+    pub fn bean() -> i32 {unimplemented!()}
+}
+
+#[test]
+fn inline_method() {
+    let mut mock = MockFoo::new();
+    mock.expect_foo()
+        .return_const(42i32);
+    assert_eq!(mock.foo(), 42);
+}
+
+#[test]
+fn inline_static_method() {
+    let ctx = MockFoo::bar_context();
+    ctx.expect()
+        .return_const(42i32);
+    assert_eq!(MockFoo::bar(), 42);
+}
+
+#[test]
+fn cold_method() {
+    let mut mock = MockFoo::new();
+    mock.expect_baz()
+        .return_const(42i32);
+    assert_eq!(mock.baz(), 42);
+}
+
+#[test]
+fn cold_static_method() {
+    let ctx = MockFoo::bean_context();
+    ctx.expect()
+        .return_const(42i32);
+    assert_eq!(MockFoo::bean(), 42);
+}

--- a/mockall/tests/automock_must_use.rs
+++ b/mockall/tests/automock_must_use.rs
@@ -1,0 +1,60 @@
+// vim: tw=80
+//! Mockall should be able to mock #[must_use] methods.  The generated code
+//! should contain #[must_use] on the mock method, but not the expect method.
+#![deny(warnings)]
+
+use mockall::*;
+
+pub struct Foo {}
+
+#[automock]
+impl Foo {
+    #[must_use]
+    pub fn bloob(&self) -> i32 {unimplemented!()}
+    #[must_use]
+    pub fn blarg() -> i32 {unimplemented!()}
+}
+
+// test that basic code generation works with must_use structs and traits.  The
+// exact output will be verified by the unit tests.
+#[must_use]
+pub struct MustUseStruct {}
+#[automock]
+impl MustUseStruct {}
+#[automock]
+#[must_use]
+pub trait MustUseTrait {}
+
+#[test]
+fn must_use_method() {
+    let mut mock = MockFoo::new();
+    mock.expect_bloob()
+        .return_const(42i32);
+    assert_eq!(42, mock.bloob());
+}
+
+#[cfg(feature = "nightly")]
+#[test]
+fn may_not_use_expectation() {
+    let mut mock = MockFoo::new();
+    // This should not produce a "must_use" warning.
+    mock.expect_bloob();
+}
+
+#[test]
+fn must_use_static_method() {
+    let ctx = MockFoo::blarg_context();
+    ctx.expect()
+        .return_const(42i32);
+    assert_eq!(MockFoo::blarg(), 42);
+}
+
+#[cfg(feature = "nightly")]
+#[test]
+fn may_not_use_static_expectation() {
+    let ctx = MockFoo::blarg_context();
+    // This should not produce a "must_use" warning.
+    ctx.expect();
+}
+
+

--- a/mockall_derive/src/lib.rs
+++ b/mockall_derive/src/lib.rs
@@ -773,6 +773,12 @@ impl<'a> AttrFormatter<'a> {
                     self.doc
                 } else if *i.as_ref().unwrap() == "async_trait" {
                     self.async_trait
+                } else if *i.as_ref().unwrap() == "inline" {
+                    // No need to inline mock functions.
+                    false
+                } else if *i.as_ref().unwrap() == "cold" {
+                    // No need for such hints on mock functions.
+                    false
                 } else if *i.as_ref().unwrap() == "instrument" {
                     // We can't usefully instrument the mock method, so just
                     // ignore this attribute.

--- a/mockall_derive/src/lib.rs
+++ b/mockall_derive/src/lib.rs
@@ -727,11 +727,11 @@ fn find_lifetimes(ty: &Type) -> HashSet<Lifetime> {
     }
 }
 
-
 struct AttrFormatter<'a>{
     attrs: &'a [Attribute],
     async_trait: bool,
     doc: bool,
+    must_use: bool,
 }
 
 impl<'a> AttrFormatter<'a> {
@@ -739,7 +739,8 @@ impl<'a> AttrFormatter<'a> {
         Self {
             attrs,
             async_trait: true,
-            doc: true
+            doc: true,
+            must_use: false,
         }
     }
 
@@ -750,6 +751,11 @@ impl<'a> AttrFormatter<'a> {
 
     fn doc(&mut self, allowed: bool) -> &mut Self {
         self.doc = allowed;
+        self
+    }
+
+    fn must_use(&mut self, allowed: bool) -> &mut Self {
+        self.must_use = allowed;
         self
     }
 
@@ -788,6 +794,8 @@ impl<'a> AttrFormatter<'a> {
                     // This shows up sometimes when mocking ffi functions.  We
                     // must not emit it on anything that isn't an ffi definition
                     false
+                } else if *i.as_ref().unwrap() == "must_use" {
+                    self.must_use
                 } else {
                     true
                 }
@@ -1329,7 +1337,7 @@ fn assert_contains(output: &str, tokens: TokenStream) {
 
 fn assert_not_contains(output: &str, tokens: TokenStream) {
     let s = tokens.to_string();
-    assert!(!output.contains(&s), "output does not contain {:?}", &s);
+    assert!(!output.contains(&s), "output contains {:?}", &s);
 }
 
 /// Various tests for overall code generation that are hard or impossible to
@@ -1366,6 +1374,17 @@ mod mock {
         assert_contains(&output, quote!(pub(crate) fn expect_baz));
         assert_contains(&output, quote!(pub(super) fn expect_bean));
         assert_contains(&output, quote!(pub(in crate::outer) fn expect_boom));
+    }
+
+    #[test]
+    fn must_use_struct() {
+        let code = "
+            #[must_use]
+            pub Foo {}
+        ";
+        let ts = proc_macro2::TokenStream::from_str(code).unwrap();
+        let output = do_mock(ts).to_string();
+        assert_contains(&output, quote!(#[must_use] pub struct MockFoo));
     }
 
     #[test]
@@ -1444,6 +1463,47 @@ mod automock {
         assert_contains(&output, quote!(pub ( crate ) fn expect_bang));
         assert_contains(&output, quote!(pub ( in super :: x ) fn bean));
         assert_contains(&output, quote!(pub ( in super :: x ) fn expect_bean));
+    }
+
+    #[test]
+    fn must_use_method() {
+        let code = "
+        impl Foo {
+            #[must_use]
+            fn foo(&self) -> i32 {42}
+        }";
+        let ts = proc_macro2::TokenStream::from_str(code).unwrap();
+        let attrs_ts = proc_macro2::TokenStream::from_str("").unwrap();
+        let output = do_automock(attrs_ts, ts).to_string();
+        assert_not_contains(&output, quote!(#[must_use] fn expect_foo));
+        assert_contains(&output, quote!(#[must_use] #[allow(dead_code)] fn foo));
+    }
+
+    #[test]
+    fn must_use_static_method() {
+        let code = "
+        impl Foo {
+            #[must_use]
+            fn foo() -> i32 {42}
+        }";
+        let ts = proc_macro2::TokenStream::from_str(code).unwrap();
+        let attrs_ts = proc_macro2::TokenStream::from_str("").unwrap();
+        let output = do_automock(attrs_ts, ts).to_string();
+        assert_not_contains(&output, quote!(#[must_use] fn expect));
+        assert_not_contains(&output, quote!(#[must_use] fn foo_context));
+        assert_contains(&output, quote!(#[must_use] #[allow(dead_code)] fn foo));
+    }
+
+    #[test]
+    fn must_use_trait() {
+        let code = "
+        #[must_use]
+        trait Foo {}
+        ";
+        let ts = proc_macro2::TokenStream::from_str(code).unwrap();
+        let attrs_ts = proc_macro2::TokenStream::from_str("").unwrap();
+        let output = do_automock(attrs_ts, ts).to_string();
+        assert_not_contains(&output, quote!(#[must_use] struct MockFoo));
     }
 
     #[test]

--- a/mockall_derive/src/mock_function.rs
+++ b/mockall_derive/src/mock_function.rs
@@ -469,7 +469,9 @@ impl MockFunction {
     // Supplying modname is an unfortunately hack.  Ideally MockFunction
     // wouldn't need to know that.
     pub fn call(&self, modname: Option<&Ident>) -> impl ToTokens {
-        let attrs = AttrFormatter::new(&self.attrs).format();
+        let attrs = AttrFormatter::new(&self.attrs)
+            .must_use(true)
+            .format();
         let call_exprs = &self.call_exprs;
         let (_, tg, _) = if self.is_method_generic() || self.is_static() {
             &self.egenerics

--- a/mockall_derive/src/mock_item_struct.rs
+++ b/mockall_derive/src/mock_item_struct.rs
@@ -245,6 +245,7 @@ impl ToTokens for MockItemStruct {
     fn to_tokens(&self, tokens: &mut TokenStream) {
         let attrs = AttrFormatter::new(&self.attrs)
             .async_trait(false)
+            .must_use(true)
             .format();
         let consts = &self.consts;
         let debug_impl = self.debug_impl();
@@ -392,7 +393,16 @@ impl MockItemTraitImpl {
 
 impl ToTokens for MockItemTraitImpl {
     fn to_tokens(&self, tokens: &mut TokenStream) {
-        let attrs = AttrFormatter::new(&self.attrs)
+        let mod_attrs = AttrFormatter::new(&self.attrs)
+            .async_trait(false)
+            .doc(false)
+            .format();
+        let struct_attrs = AttrFormatter::new(&self.attrs)
+            .async_trait(false)
+            .doc(false)
+            .must_use(false)
+            .format();
+        let impl_attrs = AttrFormatter::new(&self.attrs)
             .async_trait(false)
             .doc(false)
             .format();
@@ -408,7 +418,7 @@ impl ToTokens for MockItemTraitImpl {
         quote!(
             #[allow(non_snake_case)]
             #[allow(missing_docs)]
-            #(#attrs)*
+            #(#mod_attrs)*
             pub mod #modname {
                 use super::*;
                 #(#priv_mods)*
@@ -416,12 +426,12 @@ impl ToTokens for MockItemTraitImpl {
             #[allow(non_camel_case_types)]
             #[allow(non_snake_case)]
             #[allow(missing_docs)]
-            #(#attrs)*
+            #(#struct_attrs)*
             struct #struct_name #ig #wc
             {
                 #(#field_definitions),*
             }
-            #(#attrs)*
+            #(#impl_attrs)*
             impl #ig ::std::default::Default for #struct_name #tg #wc {
                 fn default() -> Self {
                     Self {
@@ -429,7 +439,7 @@ impl ToTokens for MockItemTraitImpl {
                     }
                 }
             }
-            #(#attrs)*
+            #(#impl_attrs)*
             impl #ig #struct_name #tg #wc {
                 /// Validate that all current expectations for all methods have
                 /// been satisfied, and discard them.

--- a/mockall_derive/src/mock_trait.rs
+++ b/mockall_derive/src/mock_trait.rs
@@ -130,7 +130,9 @@ impl MockTrait {
     // Supplying modname is an unfortunately hack.  Ideally MockTrait
     // wouldn't need to know that.
     pub fn trait_impl(&self, modname: &Ident) -> impl ToTokens {
-        let trait_impl_attrs = &self.attrs;
+        let trait_impl_attrs = AttrFormatter::new(&self.attrs)
+            .must_use(false)
+            .format();
         let impl_attrs = AttrFormatter::new(&self.attrs)
             .async_trait(false)
             .doc(false)

--- a/mockall_derive/src/mockable_struct.rs
+++ b/mockall_derive/src/mockable_struct.rs
@@ -1,6 +1,25 @@
 // vim: tw=80
-use super::*;
-use syn::parse::{Parse, ParseStream};
+use proc_macro2::{Span, TokenStream};
+use quote::{format_ident, quote};
+use syn::{
+    *,
+    parse::{Parse, ParseStream},
+    spanned::Spanned
+};
+
+use crate::{
+    Attrs,
+    compile_error,
+    deanonymize,
+    deimplify,
+    demutify,
+    deselfify,
+    deselfify_args,
+    dewhereselfify,
+    find_ident_from_path,
+    gen_mock_ident,
+    AttrFormatter,
+};
 
 /// Make any implicit lifetime parameters explicit
 fn add_lifetime_parameters(sig: &mut Signature) {
@@ -345,7 +364,15 @@ impl MockableStruct {
 impl From<(Attrs, ItemTrait)> for MockableStruct {
     fn from((attrs, item_trait): (Attrs, ItemTrait)) -> MockableStruct {
         let trait_ = attrs.substitute_trait(&item_trait);
-        let mut attrs = trait_.attrs.clone();
+        // Strip "must_use" from a trait definition.  For traits, the "must_use"
+        // should apply only when the trait is used like "impl Trait" or "dyn
+        // Trait".  So it shouldn't necessarily affect the mock struct that
+        // implements the trait.
+        let mut attrs = AttrFormatter::new(&trait_.attrs)
+            .doc(true)
+            .async_trait(true)
+            .must_use(false)
+            .format();
         attrs.push(derive_debug());
         let vis = trait_.vis.clone();
         let name = gen_mock_ident(&trait_.ident);

